src/transformers/models/splinter/modeling_splinter.py

```diff
@@ -940,3 +941,160 @@ def forward(
+)
+class SplinterForPreTraining(SplinterPreTrainedModel):
+    def __init__(self, config):
+        super().__init__(config)
+
+        self.splinter = SplinterModel(config)
+        self.splinter_qass = QuestionAwareSpanSelectionHead(config)
+        self.question_token_id = config.question_token_id
+
+        # Initialize weights and apply final processing
+        self.post_init()
+
+    @add_start_docstrings_to_model_forward(
+        SPLINTER_INPUTS_DOCSTRING.format("batch_size, num_questions, sequence_length")
+    )
+    def forward(
+        self,
+        input_ids: Optional[torch.Tensor] = None,
+        attention_mask: Optional[torch.Tensor] = None,
+        token_type_ids: Optional[torch.Tensor] = None,
+        position_ids: Optional[torch.Tensor] = None,
+        head_mask: Optional[torch.Tensor] = None,
+        inputs_embeds: Optional[torch.Tensor] = None,
+        start_positions: Optional[torch.LongTensor] = None,
+        end_positions: Optional[torch.LongTensor] = None,
+        output_attentions: Optional[bool] = None,
+        output_hidden_states: Optional[bool] = None,
+        return_dict: Optional[bool] = None,
+        question_positions: Optional[torch.LongTensor] = None,
+    ) -> Union[Tuple, SplinterForPreTrainingOutput]:
+        r"""
+        start_positions (`torch.LongTensor` of shape `(batch_size, num_questions)`, *optional*):
+            Labels for position (index) of the start of the labelled span for computing the token classification loss.
+            Positions are clamped to the length of the sequence (`sequence_length`). Position outside of the sequence
+            are not taken into account for computing the loss.
+        end_positions (`torch.LongTensor` of shape `(batch_size, num_questions)`, *optional*):
+            Labels for position (index) of the end of the labelled span for computing the token classification loss.
+            Positions are clamped to the length of the sequence (`sequence_length`). Position outside of the sequence
+            are not taken into account for computing the loss.
+        question_positions (`torch.LongTensor` of shape `(batch_size, num_questions)`, *optional*):
+            The positions of all question tokens. If given, start_logits and end_logits will be of shape `(batch_size,
+            num_questions, sequence_length)`. If None, the first question token in each sequence in the batch will be
+            the only one for which start_logits and end_logits are calculated and they will be of shape `(batch_size,
+            sequence_length)`.
+        """
+        return_dict = return_dict if return_dict is not None else self.config.use_return_dict
+
+        if question_positions is None:
+            if start_positions is not None and end_positions is not None:
+                raise TypeError(
+                    "For span selection training, question_positions must be specified to obtain the correct loss"
+                )
```

Comment:
Since we can prepare `question_positions` using `input_ids`, shouldn't we do this check after that ?

Reply:
Good point, I think our original motivation is flawed :) The assumption that there is one question token holds for question answering with Splinter, but not for recurring span selection, as different examples can have a different number of question tokens. This makes the derivation a bit more complicated. I see two options: Either we always raise `TypeError`, effectively requiring `question_positions`, or we add a small snippet for deriving it from `input_ids`. I guess one downside of the first approach is that it would be somewhat weird to require a `typing.Optional`.

Reply:
IMO the most readable code would be:

```python
if question_positions is None and start_positions is not None and end_positions is not None:
   # loss is calculated, need question positions
   raise 
elif question_positions is None and input_ids is not None:
    # input_ids are required to compute question_positions for forward pass
    raise 
elif question_positions is None and input_ids is not None
    question_positions = ...
```

Reply:
Should be good to merge once this is addressed :)

Reply:
Looks good to me!